File: E-Commerce App Project (Tabbed)/Controllers/CartViewController.m
//
//  CartViewController.m
//  E-Commerce App Project (Tabbed)
//
//  Created by Rony Banik on 7/11/18.
//  Copyright © 2018 Rony Banik. All rights reserved.
//

#import "CartViewController.h"
#import "Item.h"
#import "Items.h"
#import "SingleItemViewController.h"


@interface CartViewController ()
{
    double updatedSingleItemPrice;
    NSNumber *updatedSingleItemPriceDoubleValueToNumber;
    double stepperClickedValue;
    BOOL editClicked;
    NSIndexPath *intendedPath;
}
@end

@implementation CartViewController

@synthesize itemImage;
@synthesize itemName;
@synthesize itemPrice;
@synthesize itemQuantity;
@synthesize itemQuantityInText;
@synthesize quantityStepper;
@synthesize cartTableView;
@synthesize totalAmountLabelOutlet,totalAmountShowOutlet;
@synthesize editCartButtonOutlet,cartContinueButtonOutlet;

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    self.checkoutCart = [ShoppingCart sharedInstance];
    [cartTableView reloadData];
    if (!(self.checkoutCart.itemsInCart.count > 0)) {
        editCartButtonOutlet.enabled=NO;
        editCartButtonOutlet.tintColor = [UIColor clearColor];
        cartContinueButtonOutlet.enabled=NO;
        cartContinueButtonOutlet.tintColor=[UIColor clearColor];
        totalAmountLabelOutlet.hidden=YES;
        totalAmountShowOutlet.hidden=YES;
    }else{
        editCartButtonOutlet.enabled=YES;
        editCartButtonOutlet.tintColor = [UIColor colorNamed:@"Cornflower Blue"];
        cartContinueButtonOutlet.enabled=YES;
        cartContinueButtonOutlet.tintColor=[UIColor colorNamed:@"Cornflower Blue"];
        totalAmountLabelOutlet.hidden=NO;
        totalAmountShowOutlet.hidden=NO;
    }
}
- (void)viewDidAppear:(BOOL)animated{
    [cartTableView reloadData];
    if (!(self.checkoutCart.itemsInCart.count > 0)) {
        editCartButtonOutlet.enabled=NO;
        editCartButtonOutlet.tintColor = [UIColor clearColor];
        cartContinueButtonOutlet.enabled=NO;
        cartContinueButtonOutlet.tintColor=[UIColor clearColor];
        totalAmountLabelOutlet.hidden=YES;
        totalAmountShowOutlet.hidden=YES;
    }else{
        editCartButtonOutlet.enabled=YES;
        editCartButtonOutlet.tintColor = [UIColor colorNamed:@"Cornflower Blue"];
        cartContinueButtonOutlet.enabled=YES;
        cartContinueButtonOutlet.tintColor=[UIColor colorNamed:@"Cornflower Blue"];
        totalAmountLabelOutlet.hidden=NO;
        totalAmountShowOutlet.hidden=NO;
    }
}
- (void)viewWillAppear:(BOOL)animated{
    [cartTableView reloadData];
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    if (indexPath.section==0) {
        Item *itemObject = self.checkoutCart.itemsInCart[indexPath.row];
        intendedPath=indexPath;
        UITableViewCell *cellIdentifier = [tableView dequeueReusableCellWithIdentifier:@"cartTableViewCellIdentifier"];
       
        itemImage = (UIImageView *)[cellIdentifier viewWithTag:701];
        itemName = (UILabel *)[cellIdentifier viewWithTag:702];
        itemPrice = (UILabel *)[cellIdentifier viewWithTag:703];
        itemQuantity = (UILabel *)[cellIdentifier viewWithTag:704];
        quantityStepper = (UIStepper *)[cellIdentifier viewWithTag:705];
        
        NSURL *url = [NSURL URLWithString:itemObject.photoURL];
        NSData *data = [NSData dataWithContentsOfURL:url];

        itemImage.image = [UIImage imageWithData:data];
        itemName.text = itemObject.itemName;
        quantityStepper.value=itemObject.cartAddedQuantity;
        
        updatedSingleItemPrice = (itemObject.cartAddedQuantity*itemObject.price);
        updatedSingleItemPriceDoubleValueToNumber = [NSNumber numberWithDouble:updatedSingleItemPrice];
        itemPrice.text = [self showPrice:updatedSingleItemPriceDoubleValueToNumber];
        
        itemQuantityInText = [[NSNumber numberWithDouble:itemObject.cartAddedQuantity] stringValue];
        itemQuantity.text = itemQuantityInText;
        totalAmountShowOutlet.text = [self showPrice:[self.checkoutCart total]];
        
        return cellIdentifier;
    }
    return nil;
}

- (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return (section == 0)? self.checkoutCart.itemsInCart.count :1;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath{
    CGFloat heightForRow;
    heightForRow = 100;
    return heightForRow;
}

- (IBAction)stepperClicked:(UIStepper *)sender
            {
    double stepperValue = [sender value];
    CGPoint buttonPosition = [sender convertPoint:CGPointZero toView:cartTableView];
    NSIndexPath *rowIndexPath = [cartTableView indexPathForRowAtPoint:buttonPosition];
    [self returnStepperValue:rowIndexPath withStepperValue:stepperValue];
}
-(void)returnStepperValue:(NSIndexPath *)indexPath
         withStepperValue:(double)stepperValue{
    Item *itemObject = self.checkoutCart.itemsInCart[indexPath.row];
    itemObject.cartAddedQuantity = stepperValue;
    [cartTableView reloadData];
}

- (IBAction)editCartButton:(id)sender {
    if(editClicked == NO){
        [cartTableView setEditing:YES animated:YES];
        editClicked = YES;
        [cartTableView reloadData];
    }else{
        [cartTableView setEditing:NO animated:YES];
        editClicked = NO;
        [cartTableView reloadData];
    }
}

- (IBAction)CartContinueButton:(id)sender {
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    if (![defaults boolForKey:@"SeesionUserLoggedIN"]) {
        [self performSegueWithIdentifier:@"cartToLogInSegue" sender:nil];
    }else{
        [self performSegueWithIdentifier:@"checkOutSegue" sender:nil];
    }
}
- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath{
    if (editingStyle==UITableViewCellEditingStyleDelete) {
        
        Item *itemObject = self.checkoutCart.itemsInCart[indexPath.row];
        [self.checkoutCart removeFromCart:itemObject];
        totalAmountShowOutlet.text=[[self.checkoutCart total] stringValue];
        [tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath] withRowAnimation:UITableViewCellEditingStyleDelete];
    }
}
-(NSString *)showPrice:(NSNumber *)price{
    NSString *priceString = [NSString stringWithFormat:@"$%@",price];
    return priceString;
}
@end

File: E-Commerce App Project (Tabbed)/Controllers/CartViewController.swift
//
//  CartViewController.swift
//  E-Commerce App Project (Tabbed)
//
//  Converted to Swift with MVVM pattern
//

import UIKit

class CartViewController: UIViewController, UITableViewDataSource, UITableViewDelegate {
    
    // MARK: - ViewModel
    private let viewModel = CartViewModel()
    
    // MARK: - IBOutlets
    @IBOutlet var cartTableView: UITableView!
    @IBOutlet var totalAmountLabelOutlet: UILabel!
    @IBOutlet var totalAmountShowOutlet: UILabel!
    @IBOutlet var editCartButtonOutlet: UIBarButtonItem!
    @IBOutlet var cartContinueButtonOutlet: UIBarButtonItem!
    
    private var editClicked: Bool = false
    
    override func viewDidLoad() {
        super.viewDidLoad()
        bindViewModel()
        viewModel.refreshCart()
    }
    
    private func bindViewModel() {
        viewModel.isCartEmpty.bind { [weak self] isEmpty in
            guard let self = self else { return }
            self.updateUIForCartState(isEmpty: isEmpty)
        }
        
        viewModel.totalAmount.bind { [weak self] total in
            self?.totalAmountShowOutlet.text = total
        }
    }
    
    private func updateUIForCartState(isEmpty: Bool) {
        editCartButtonOutlet.isEnabled = !isEmpty
        editCartButtonOutlet.tintColor = isEmpty ? .clear : UIColor(named: "Cornflower Blue")
        cartContinueButtonOutlet.isEnabled = !isEmpty
        cartContinueButtonOutlet.tintColor = isEmpty ? .clear : UIColor(named: "Cornflower Blue")
        totalAmountLabelOutlet.isHidden = isEmpty
        totalAmountShowOutlet.isHidden = isEmpty
        cartTableView.reloadData()
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        viewModel.refreshCart()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        viewModel.refreshCart()
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        if indexPath.section == 0 {
            let itemObject = viewModel.item(at: indexPath.row)
            let cell = tableView.dequeueReusableCell(withIdentifier: "cartTableViewCellIdentifier")!
            
            let itemImage = cell.viewWithTag(701) as? UIImageView
            let itemName = cell.viewWithTag(702) as? UILabel
            let itemPrice = cell.viewWithTag(703) as? UILabel
            let itemQuantity = cell.viewWithTag(704) as? UILabel
            let quantityStepper = cell.viewWithTag(705) as? UIStepper
            
            if let url = URL(string: itemObject.photoURL),
               let data = try? Data(contentsOf: url) {
                itemImage?.image = UIImage(data: data)
            }
            itemName?.text = itemObject.itemName
            quantityStepper?.value = viewModel.itemQuantity(at: indexPath.row)
            itemPrice?.text = viewModel.itemPrice(at: indexPath.row)
            itemQuantity?.text = "\(Int(viewModel.itemQuantity(at: indexPath.row)))"
            
            return cell
        }
        return UITableViewCell()
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return section == 0 ? viewModel.numberOfItems() : 1
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 100
    }
    
    @IBAction func stepperClicked(_ sender: UIStepper) {
        let stepperValue = sender.value
        let buttonPosition = sender.convert(CGPoint.zero, to: cartTableView)
        if let rowIndexPath = cartTableView.indexPathForRow(at: buttonPosition) {
            viewModel.updateQuantity(at: rowIndexPath.row, quantity: stepperValue)
            cartTableView.reloadData()
        }
    }
    
    @IBAction func editCartButton(_ sender: Any) {
        editClicked = !editClicked
        cartTableView.setEditing(editClicked, animated: true)
        cartTableView.reloadData()
    }
    
    @IBAction func CartContinueButton(_ sender: Any) {
        if viewModel.shouldNavigateToCheckout() {
            performSegue(withIdentifier: "checkOutSegue", sender: nil)
        } else {
            performSegue(withIdentifier: "cartToLogInSegue", sender: nil)
        }
    }
    
    func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCell.EditingStyle, forRowAt indexPath: IndexPath) {
        if editingStyle == .delete {
            viewModel.removeItem(at: indexPath.row)
            tableView.deleteRows(at: [indexPath], with: .automatic)
        }
    }
}
